### PR TITLE
Fixes from inference track

### LINKED ISSUE #1093: Fixes from inference track
## Proposal
1. Check that all the transforms have `test_save_load` test, add this test where necessary 
2. Fix tests in `test_models/test_inference` folder
3. Add tests for saving/loading for `HierarchicalPipeline` + `BaseReconciliator`

## Expected results
1. All transforms have `test_save_load` test
2. All tests from `test_models/test_inference` folder passed
3. `HierarchicalPipeline` + `BaseReconciliator` has tests for saving/loading

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look #1093.

1. We should discuss is suggested solution for `HierarchicalPipeline` ok.
2. We should discuss how are we going to fix old NNs (they broke).

## Closing issues
Closes #1093.

### CODE REVIEW

File: etna/pipeline/hierarchical_pipeline.py [hunk 1]
```diff
@@ -154,3 +155,23 @@ def _forecast_prediction_interval(
...
+        ts:
+            TSDataset to set into loaded pipeline.
+
+        Returns
+        -------
+        :
+            Loaded object.
+        """
+        obj = super().load(path=path)
+        obj.ts = obj.reconciliator.aggregate(ts=ts)
```
Comment (alex-hse-repository) [inline etna/pipeline/hierarchical_pipeline.py:176]: We also need to save `self._fit_ts` -- we need it for prediction intervals

File: tests/test_pipeline/test_hierarchical_pipeline.py [hunk 2]
```diff
@@ -260,3 +264,33 @@ def test_interval_metrics(product_level_constant_hierarchical_ts, metric_type, r
...
+            [DateFlagsTransform(), LagTransform(in_column="target", lags=[1])],
+        ),
+        (NaiveModel(), []),
+        (ProphetModel(), []),
+    ],
+)
+def test_save_load(model, transforms, reconciliator, product_level_constant_hierarchical_ts):
+    horizon = 1
+    pipeline = HierarchicalPipeline(reconciliator=reconciliator, model=model, transforms=transforms, horizon=horizon)
+    assert_pipeline_equals_loaded_original(pipeline=pipeline, ts=product_level_constant_hierarchical_ts)
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_hierarchical_pipeline.py:296]: We should create separate method for this pipeline as it saves 2 datasets inside: raw and aggregated

File: tests/test_transforms/test_decomposition/test_change_points_based/test_level.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 from etna.transforms import ChangePointsLevelTransform
 from etna.transforms.decomposition.change_points_based.change_points_models import RupturesChangePointsModel
 from etna.transforms.decomposition.change_points_based.per_interval_models import MeanPerIntervalModel
+from tests.test_transforms.utils import assert_transformation_equals_loaded_original
```
Comment (alex-hse-repository) [inline tests/test_transforms/test_decomposition/test_change_points_based/test_level.py:12]: Do all the other transforms have this test? It seemed to me that there were more

File: etna/models/nn/utils.py [hunk 4]
```diff
@@ -236,7 +236,33 @@ def fit(self, ts: TSDataset):
...
+    def _get_first_prediction_timestamp(self, ts: TSDataset, horizon: int) -> pd.Timestamp:
+        return ts.index[-horizon]
+
+    def _is_in_sample_prediction(self, ts: TSDataset, horizon: int) -> bool:
+        first_prediction_timestamp = self._get_first_prediction_timestamp(ts=ts, horizon=horizon)
+        return first_prediction_timestamp <= self._last_train_timestamp
+
+    def _is_prediction_with_gap(self, ts: TSDataset, horizon: int) -> bool:
+        first_prediction_timestamp = self._get_first_prediction_timestamp(ts=ts, horizon=horizon)
+        first_timestamp_after_train = pd.date_range(self._last_train_timestamp, periods=2, freq=self._freq)[-1]
```
Comment (alex-hse-repository) [inline etna/models/nn/utils.py:248]: May be we can solve it with timedelta?

File: etna/models/nn/utils.py [hunk 5]
```diff
@@ -236,7 +236,33 @@ def fit(self, ts: TSDataset):
...
+        return first_prediction_timestamp > first_timestamp_after_train
+
     def _make_target_prediction(self, ts: TSDataset, horizon: int) -> Tuple[TSDataset, DataLoader]:
+        if self._is_in_sample_prediction(ts=ts, horizon=horizon):
+            raise NotImplementedError(
+                "It is not possible to make in-sample predictions with DeepAR model! "
+                "In-sample predictions aren't supported by current implementation."
+            )
+        elif self._is_prediction_with_gap(ts=ts, horizon=horizon):
```
Comment (alex-hse-repository) [inline etna/models/nn/utils.py:257]: As i understand we should be able to forecast with gap

File: tests/test_models/test_inference/test_forecast.py [hunk 6]
```diff
@@ -459,36 +479,54 @@ def _test_forecast_out_sample_prefix(ts, model, transforms, full_prediction_size
             (DeadlineMovingAverageModel(window=1), []),
             (BATSModel(use_trend=True), []),
             (TBATSModel(use_trend=True), []),
+            (RNNModel(input_size=1, encoder_length=7, decoder_length=7, trainer_params=dict(max_epochs=1)), []),
+        ],
+    )
+    def test_forecast_out_sample_prefix(self, model, transforms, example_tsds):
+        self._test_forecast_out_sample_prefix(example_tsds, model, transforms)
+
+    # if we ignore this warning there is assertion error
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:488]: May you create an issue

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_inference/test_forecast.py [hunk 7]
```diff
@@ -459,36 +479,54 @@ def _test_forecast_out_sample_prefix(ts, model, transforms, full_prediction_size
...
+        [
             (
-                DeepARModel(max_epochs=5, learning_rate=[0.01]),
-                [
-                    PytorchForecastingTransform(
+                DeepARModel(
+                    dataset_builder=PytorchForecastingDatasetBuilder(
                         max_encoder_length=5,
                         max_prediction_length=5,
+                        min_prediction_length=1,
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:501]: Remove this + look in PTF docs, may be we can fix this bumping the version

File: etna/models/nn/utils.py (hunk 4 above)
Comment (Mr-Geekman) [reply etna/models/nn/utils.py:248]: I don't think it is a good idea. Frequency works more subtle than simple timedelta, as I understand.

File: etna/models/nn/utils.py (hunk 5 above)
Comment (Mr-Geekman) [reply etna/models/nn/utils.py:257]: In theory, yes. On practice it isn't currently working. I haven't studied it very thoroughly, but tests just doesn't work as expected. So I decided to add this fence.

File: tests/test_models/test_inference/test_forecast.py (hunk 6 above)
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:488]: Yes, I think we should investigate why old NNs isn't working as expected. But I don't think we should make the fix inside this task.

File: tests/test_pipeline/test_hierarchical_pipeline.py [hunk 8]
```diff
@@ -266,12 +269,71 @@ def test_interval_metrics(product_level_constant_hierarchical_ts, metric_type, r
...
+    reconciliator = BottomUpReconciliator(target_level="market", source_level="product")
+    pipeline = HierarchicalPipeline(reconciliator=reconciliator, model=model, transforms=[], horizon=1)
+    dir_path = pathlib.Path(tmp_path)
+    path = dir_path / "dummy.zip"
+    pipeline.fit(ts)
+
+    def check_no_fit_ts(path):
+        assert not hasattr(pipeline, "_fit_ts")
+
+    mocked_load.side_effect = check_no_fit_ts
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_hierarchical_pipeline.py:285]: mocked_save?

File: etna/pipeline/hierarchical_pipeline.py [hunk 9]
```diff
@@ -156,6 +156,25 @@ def _forecast_prediction_interval(
...
+        """Save the object.
+
+        Parameters
+        ----------
+        path:
+            Path to save object to.
+        """
+        fit_ts = self._fit_ts
+
+        try:
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/pipeline/hierarchical_pipeline.py:169]: What is your alternative? I wanted to make action inside `finally` block even if there is some exception. That is why I did it like this.

**Pull request merged**